File: product_sum/product_sum.py
# Complexity O(n) n - number of all arrays elements,
# Space O(d) d - maximal depth of array
def productSum(array, d = 1):
    summa = 0
    for n in array:
        if type(n) is list:
            summa += productSum(n, d + 1)
        else:
            summa += n
    return summa * d


if __name__ == "__main__":
    test = [5, 2, [7, -1], 3, [6, [-13, 8], 4]]
    print(12, "vs", productSum(test))

File: product_sum/productSum.go
package main

import "fmt"

func productSum(array []interface{}, depth int) int {
	var res int

	for _, el := range array {
		if v, ok := el.(int); ok {
			res += v
		} else {
			res += productSum(el.([]interface{}), depth + 1)
		}
	}
	return res * depth
}

func main() {
	var array = []interface{}{5, 2, []interface{}{7, -1}, 3, []interface{}{6, []interface{}{-13, 8}, 4}}
	fmt.Println(12, "vs", productSum(array, 1))
}
